cli/lsp/analysis.rs

```diff
@@ -948,6 +984,136 @@ fn prepend_whitespace(content: String, line_content: Option<String>) -> String {
   }
 }
 
+/// Get LSP range from the provided SWC start and end locations.
+fn get_range_from_loc(start: &Loc, end: &Loc) -> lsp::Range {
+  lsp::Range {
+    start: lsp::Position {
+      line: (start.line - 1) as u32,
+      character: (start.col_display + 1) as u32,
+    },
+    end: lsp::Position {
+      line: (end.line - 1) as u32,
+      character: (end.col_display - 1) as u32,
+    },
+  }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TextRange {
+  pub range: lsp::Range,
+  pub text: String,
+}
+
+impl TextRange {
+  fn includes(&self, position: &lsp::Position) -> bool {
+    (position.line > self.range.start.line
+      || position.line == self.range.start.line
+        && position.character >= self.range.start.character)
+      && (position.line < self.range.end.line
+        || position.line == self.range.end.line
+          && position.character <= self.range.end.character)
+  }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct ImportRanges(Vec<TextRange>);
+
+impl ImportRanges {
+  pub fn contains(&self, position: &lsp::Position) -> Option<TextRange> {
+    self.0.iter().find(|r| r.includes(position)).cloned()
+  }
+}
+
+struct ImportRangeCollector {
+  import_ranges: ImportRanges,
+  source_map: Rc<SourceMap>,
+}
+
+impl ImportRangeCollector {
+  pub fn new(source_map: Rc<SourceMap>) -> Self {
+    Self {
+      import_ranges: ImportRanges::default(),
+      source_map,
+    }
+  }
+
+  pub fn take(self) -> ImportRanges {
+    self.import_ranges
+  }
+}
+
+impl Visit for ImportRangeCollector {
+  fn visit_import_decl(
+    &mut self,
+    node: &swc_ast::ImportDecl,
+    _parent: &dyn Node,
+  ) {
+    let start = self.source_map.lookup_char_pos(node.src.span.lo);
+    let end = self.source_map.lookup_char_pos(node.src.span.hi);
```

Comment:
Yeah, upon reflection, I think you are right, better to just collect the spans at this point.

Reply:
Actually, looking at this it is far more complicated than I expected. So the challenges are:

- We can't hold onto a SourceMap in the documents or language server, because swc source maps are not thread safe and the documents structure needs to be access by various threads, which we can't predict.
- We have the line index that we use to convert from LSP to tsc, but in looking at how we might use that to remap an swc span to a range without the source map, I realised that we don't actually convert any swc positions to lsp ones.
- We had all sorts of issues with dealing with 16-bit encoding of strings and figuring the right positions in documents with the line index, so I really don't want to go down that rabbit hole right now with swc to lsp, I would much rather trust the swc source map, as that we have had a lot of experience with.

So while it isn't efficient, we basically never hold onto a swc pos/loc in the language server at this point, we always convert to "native" lsp (like in the dependency analysis).  So while it seems like a good idea, I think we either would want to go back and do it as a wholesale change in the other areas as well as identify the right pattern for converting from lsp <-> swc without using the source map.

Reply:
> I think we either would want to go back and do it as a wholesale change in the other areas as well as identify the right pattern for converting from lsp <-> swc without using the source map.

Yup, agreed! I will open an issue.